From 622241d7b15f23cd3f94d5f92137986acb746cde Mon Sep 17 00:00:00 2001
From: Michael Heimpold <michael.heimpold@chargebyte.com>
Date: Tue, 19 May 2026 14:19:05 +0200
Subject: [PATCH] fix(PN532TokenProvider): rework low-level frame sending and
 receiving (#2095)

* config(bringup): add a bringup setup for PN532TokenProvider

Signed-off-by: Michael Heimpold <mhei@heimpold.de>

* framework(serial): fix initialization and shutdown

Using zero as initial value is bad since zero is a valid fd.
Let's use -1 instead and also check for this value during shutdown.

And since we have virtual method, we have to declare the destructor virtual, too.

Signed-off-by: Michael Heimpold <mhei@heimpold.de>

* framework(serial): add hexdump methods

Signed-off-by: Michael Heimpold <mhei@heimpold.de>

* PN532TokenProvider: rework low-level frame sending and receiving

While testing this module on chargebyte platforms, it did not work at all.
I did not manage to understand the frame parsing and analyzing code,
so I rewrote it.
This code was test on a native UART on Charge SOM, but also works
with a USB to serial adapter on PC.

Signed-off-by: Michael Heimpold <mhei@heimpold.de>

---------

Signed-off-by: Michael Heimpold <mhei@heimpold.de>
---
 config/bringup/CMakeLists.txt                 |   1 +
 .../config-bringup-pn532-token-provider.yaml  |  16 +
 .../framework/include/utils/serial.hpp        |  11 +-
 lib/everest/framework/lib/serial.cpp          |  19 +-
 .../main/auth_token_providerImpl.cpp          |   6 +-
 .../pn532_serial/PN532Serial.cpp              | 344 ++++++++----------
 .../pn532_serial/PN532Serial.h                |  28 +-
 7 files changed, 212 insertions(+), 213 deletions(-)
 create mode 100644 config/bringup/config-bringup-pn532-token-provider.yaml

diff --git a/config/bringup/CMakeLists.txt b/config/bringup/CMakeLists.txt
index 840bc8bdbc4b..759d3e56f7ab 100644
--- a/config/bringup/CMakeLists.txt
+++ b/config/bringup/CMakeLists.txt
@@ -17,6 +17,7 @@ generate_config_tmux_run_script(CONFIG bringup-dold-rn5893)
 generate_config_tmux_run_script(CONFIG bringup-huawei)
 generate_config_tmux_run_script(CONFIG bringup-isolation-monitor-sil)
 generate_config_tmux_run_script(CONFIG bringup-nxpnfcfrontend-token-provider)
+generate_config_tmux_run_script(CONFIG bringup-pn532-token-provider)
 generate_config_tmux_run_script(CONFIG bringup-system)
 generate_config_tmux_run_script(CONFIG bringup-uugreen)
 generate_config_tmux_run_script(CONFIG bringup-yetidriver)
diff --git a/config/bringup/config-bringup-pn532-token-provider.yaml b/config/bringup/config-bringup-pn532-token-provider.yaml
new file mode 100644
index 000000000000..c800748d7cd1
--- /dev/null
+++ b/config/bringup/config-bringup-pn532-token-provider.yaml
@@ -0,0 +1,16 @@
+active_modules:
+  tokenprovider:
+    module: PN532TokenProvider
+    config_implementation:
+      main:
+        serial_port: /dev/ttyUSB0
+        baud_rate: 115200
+        debug: true
+        read_timeout: 2
+  cli:
+    module: BUTokenProvider
+    standalone: true
+    connections:
+      token_provider:
+        - module_id: tokenprovider
+          implementation_id: main
diff --git a/lib/everest/framework/include/utils/serial.hpp b/lib/everest/framework/include/utils/serial.hpp
index 33f5811e9452..c25a80009649 100644
--- a/lib/everest/framework/include/utils/serial.hpp
+++ b/lib/everest/framework/include/utils/serial.hpp
@@ -6,22 +6,27 @@
 #include <cstddef>
 #include <cstdio>
 #include <stdint.h>
+#include <string>
 #include <termios.h>
+#include <vector>
 
 namespace Everest {
 class Serial {
 
 public:
     Serial();
-    ~Serial();
+    virtual ~Serial();
 
     bool openDevice(const char* device, int baud);
 
     virtual void run() = 0;
 
+    virtual std::string hexdump(const std::uint8_t* const msg, int msg_len) const;
+    virtual std::string hexdump(const std::vector<std::uint8_t>& msg) const;
+
 protected:
-    int fd;
-    int baud;
+    int fd{-1};
+    int baud{0};
 
 private:
     // Serial interface
diff --git a/lib/everest/framework/lib/serial.cpp b/lib/everest/framework/lib/serial.cpp
index 39f06071324e..3f82a4858eaf 100644
--- a/lib/everest/framework/lib/serial.cpp
+++ b/lib/everest/framework/lib/serial.cpp
@@ -11,13 +11,11 @@
 
 namespace Everest {
 Serial::Serial() {
-    fd = 0;
-    baud = 0;
     cobsDecodeReset();
 }
 
 Serial::~Serial() {
-    if (fd) {
+    if (fd != -1) {
         close(fd);
     }
 }
@@ -163,4 +161,19 @@ void Serial::cobsDecodeByte(uint8_t byte) {
     block--;
 }
 
+std::string Serial::hexdump(const std::uint8_t* const msg, int msg_len) const {
+    std::stringstream ss;
+    for (int i = 0; i < msg_len; i++) {
+        if (i > 0) {
+            ss << " ";
+        }
+        ss << "0x" << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(msg[i]);
+    }
+    return ss.str();
+}
+
+std::string Serial::hexdump(const std::vector<std::uint8_t>& msg) const {
+    return hexdump(msg.data(), msg.size());
+}
+
 } // namespace Everest
diff --git a/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/main/auth_token_providerImpl.cpp b/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/main/auth_token_providerImpl.cpp
index 084b4167ef97..f6e6d9003b27 100644
--- a/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/main/auth_token_providerImpl.cpp
+++ b/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/main/auth_token_providerImpl.cpp
@@ -40,7 +40,7 @@ void auth_token_providerImpl::ready() {
     }
     auto configure_sam = configure_sam_future.get();
     if (configure_sam) {
-        EVLOG_debug << "Configured SAM" << std::endl;
+        EVLOG_debug << "Configured SAM";
     }
 
     auto firmware_version_future = serial.getFirmwareVersion();
@@ -51,9 +51,7 @@ void auth_token_providerImpl::ready() {
     auto firmware_version = firmware_version_future.get();
     if (firmware_version.valid) {
         std::shared_ptr<FirmwareVersion> fv = std::dynamic_pointer_cast<FirmwareVersion>(firmware_version.message);
-        if (config.debug) {
-            EVLOG_info << "PN532 firmware version: " << std::to_string(fv->ver) << "." << std::to_string(fv->rev);
-        }
+        EVLOG_info << "PN532 firmware version: " << std::to_string(fv->ver) << "." << std::to_string(fv->rev);
     }
 
     while (true) {
diff --git a/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.cpp b/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.cpp
index e435248a0f85..5f15d1e9c5fc 100644
--- a/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.cpp
+++ b/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.cpp
@@ -5,6 +5,7 @@
 #include <fcntl.h>
 #include <fstream>
 #include <iostream>
+#include <string.h>
 #include <string>
 #include <thread>
 #include <unistd.h>
@@ -31,188 +32,159 @@ constexpr uint8_t COMMAND_SAM_CONFIGURATION_RESPONSE = 0x15;
 constexpr uint8_t COMMAND_IN_LIST_PASSIVE_TARGET = 0x4a;
 constexpr uint8_t COMMAND_IN_LIST_PASSIVE_TARGET_RESPONSE = 0x4b;
 
-PN532Serial::PN532Serial() {
-}
-
-PN532Serial::~PN532Serial() {
-}
-
 void PN532Serial::run() {
     readThreadHandle = std::thread(&PN532Serial::readThread, this);
 }
 
-void PN532Serial::resetDataRead() {
-    start_of_packet = 0;
-    packet_length = 0;
-    preamble_start_seen = false;
-    preamble_seen = false;
-    first_data = true;
-    data_length_checksum_valid = false;
-    tfi = 0;
-    command_code = 0;
-    data.clear();
-}
-
 void PN532Serial::readThread() {
     uint8_t buf[2048];
-
-    resetDataRead();
+    size_t n = 0;
+    ssize_t first_start_code_offset;
 
     while (true) {
         if (readThreadHandle.shouldExit()) {
             break;
         }
-        auto n = read(fd, buf, sizeof buf);
-        if (n == 0) {
+        auto c = read(fd, &buf[n], sizeof(buf) - n);
+        if (c == -1) {
+            EVLOG_error << "Read failed: " << strerror(errno) << " (" << errno << ")";
+            break;
+        }
+        if (c == 0) {
             continue;
         }
 
         if (this->debug) {
-            std::stringstream data_stream;
-            for (ssize_t i = 0; i < n; i++) {
-                data_stream << "0x" << std::setfill('0') << std::setw(2) << std::right << std::hex << (int)buf[i]
-                            << " ";
-            }
+            EVLOG_info << "Received: " << hexdump(buf + n, c);
+        }
+
+        n += c;
+
+    restart_buffer_analysis:
+        EVLOG_verbose << "buffer content (size: " << n << "): " << hexdump(buf, n);
 
-            EVLOG_info << "Received bytes: " << data_stream.str();
+        // packet minimum size is 6 byte for ACK or NACK, other frames are larger
+        if (n < 6) {
+            continue;
         }
-        for (ssize_t i = 0; i < n; i++) {
-            if (!preamble_seen) {
-                if (preamble_start_seen && buf[i] == 0xff) {
-                    preamble_seen = true;
-                    start_of_packet = i + 1;
-                    if (start_of_packet + 1 >= n) {
-                        if (this->debug) {
-                            EVLOG_warning << "Packet is not long enough to be parsed";
-                        }
-                        continue;
-                    } else {
-                        packet_length = buf[start_of_packet];
-                        auto packet_length_checksum = buf[start_of_packet + 1];
-                        auto checksum = (packet_length + packet_length_checksum) & 0x00ff;
-                        if (checksum == 0) {
-                            // data length checkum is valid
-                            data_length_checksum_valid = true;
-                        } else {
-                            // can be a valid ACK frame
-                        }
-                    }
-                    break;
-                } else {
-                    preamble_start_seen = false;
+
+        first_start_code_offset = -1;
+
+        // we cannot assume that we receive the whole response in one read call
+        // because of different driver/interfaces, so we have to scan for well-known
+        // patterns or check whether it looks like a valid frame after a start code
+        // pattern; if we find one, we can delete it from the buffer and move the
+        // left-over buffer content to the front, then we must restart analyzing
+        // to be sure to handle any following frame immediately and don't wait
+        // until the next read call runs/returns
+        for (ssize_t i = 0; i <= n - 6; ++i) {
+            // check for ACK or NACKs
+            if (memcmp(&buf[i], ACK_FRAME.data(), ACK_FRAME.size()) == 0) {
+                EVLOG_debug << "Received ACK";
+                n = n - i - ACK_FRAME.size();
+                memmove(buf, buf + i + ACK_FRAME.size(), n);
+                goto restart_buffer_analysis;
+            }
+            if (memcmp(&buf[i], NACK_FRAME.data(), NACK_FRAME.size()) == 0) {
+                EVLOG_debug << "Received NACK";
+                n = n - i - NACK_FRAME.size();
+                memmove(buf, buf + i + NACK_FRAME.size(), n);
+                goto restart_buffer_analysis;
+            }
+
+            // check for start_code
+            if (memcmp(&buf[i], START_CODE.data(), START_CODE.size()) == 0) {
+                if (first_start_code_offset == -1) {
+                    first_start_code_offset = i;
                 }
 
-                if (!preamble_start_seen && buf[i] == 0x00) {
-                    preamble_start_seen = true;
+                auto len = buf[i + START_CODE.size()];
+                auto lcs = buf[i + START_CODE.size() + 1];
+
+                if (((len + lcs) & 0xff) != 0) {
+                    EVLOG_verbose << "package length checksum mismatch";
+                    continue;
                 }
-            } else {
-                data.push_back(buf[i]);
-            }
-        }
 
-        if (packet_length == 0) {
-            if (start_of_packet + 1 < n) {
-                if (buf[start_of_packet + 1] == 0xff) {
-                    // ACK FRAME
-                    // TODO keep track of acks
+                // offset + start code size + len and lcs byte + data length + dcs and postamble
+                if (i + START_CODE.size() + 2 + len + 2 > n) {
+                    EVLOG_verbose << "frame still too short or length bogus";
+                    continue;
                 }
-            } else {
-                if (this->debug) {
-                    EVLOG_warning << "Packet of length 0 received that is not a ACK Frame.";
+
+                // accesses beyond i are safe below since we ensure that above
+                auto tfi = buf[i + START_CODE.size() + 2];
+                if (tfi != PN532_TO_HOST) {
+                    EVLOG_verbose << "frame is not a response";
+                    continue;
                 }
-                resetDataRead();
-            }
-        } else {
-            if (!data_length_checksum_valid) {
-                if (this->debug) {
-                    EVLOG_warning << "Data length checksum invalid";
+
+                if (buf[i + START_CODE.size() + 2 + len + 1] != 0x00) {
+                    EVLOG_verbose << "postamble check failed";
+                    continue;
                 }
-                resetDataRead();
-                continue;
-            }
 
-            // normal packet
-            if (first_data) {
-                tfi = buf[start_of_packet + 2];
-                command_code = buf[start_of_packet + 3];
-                first_data = false;
-                for (ssize_t i = start_of_packet + 4; i < n; i++) {
-                    data.push_back(buf[i]);
+                auto dcs_compl = 0;
+                for (size_t j = 0; j < len; ++j) {
+                    dcs_compl += buf[i + START_CODE.size() + 2 + j];
+                }
+                auto dcs = buf[i + START_CODE.size() + 2 + len];
+                if (((dcs_compl + dcs) & 0xff) != 0) {
+                    EVLOG_verbose << "package data checksum does not match";
+                    continue;
                 }
-            }
 
-            if (data.size() < packet_length) {
-                // not enough bytes received for a complete message yet
-                continue;
-            } else {
+                // all checks so far indicate that frame/packet looks good
+                command_code = buf[i + START_CODE.size() + 2 + 1];
+                data = std::vector<uint8_t>(&buf[i + START_CODE.size() + 2 + 1 + 1],
+                                            &buf[i + START_CODE.size() + 2 + len]);
                 parseData();
-            }
-        }
-        resetDataRead();
-    }
-}
 
-void PN532Serial::parseData() {
-    if (data.back() == 0x00 && data.size() >= 2) {
-        // last byte is 0x00 (postamble) removing it
-        data.pop_back();
-        auto packet_data_checksum = data.back();
-        data.pop_back();
-
-        uint8_t sum = tfi + command_code + packet_data_checksum;
-        for (auto element : data) {
-            sum += element;
-        }
-        if (this->debug) {
-            EVLOG_info << "----- Parsing data -----";
-            std::stringstream data_stream;
-            EVLOG_info << std::right << std::hex << "Packet data checksum: 0x" << std::setfill('0') << std::setw(2)
-                       << (int)packet_data_checksum << " checksum result: 0x" << std::setfill('0') << std::setw(2)
-                       << (int)sum;
-            for (auto& element : data) {
-                data_stream << "0x" << std::setfill('0') << std::setw(2) << std::right << std::hex << (int)element
-                            << " ";
+                // discard processed data, the processed size is:
+                // offset (incl. preamble) + start code size + len and lcs byte + data length + dcs and postamble
+                auto processed_size = i + START_CODE.size() + 1 + 1 + data.size() + 1 + 1;
+                n = n - processed_size;
+                memmove(buf, buf + processed_size, n);
+                goto restart_buffer_analysis;
             }
-            data_stream << "(length: " << std::dec << data.size() << ")";
-
-            EVLOG_info << "Raw data: " << data_stream.str();
         }
 
-        if (sum == 0x00) {
-            if (this->debug) {
-                EVLOG_info << "Packet data checksum valid, command code: 0x" << std::setfill('0') << std::setw(2)
-                           << std::right << std::hex << (int)command_code;
-            }
-
-            if (command_code == COMMAND_GET_FIRMWARE_VERSION_RESPONSE && this->get_firmware_version_promise) {
-                PN532Response response;
-                if (data.size() == 4) {
-                    response.valid = true;
-                    response.message = std::make_shared<FirmwareVersion>(
-                        data.at(FIRMWARE_VERSION_IC_POS), data.at(FIRMWARE_VERSION_VER_POS),
-                        data.at(FIRMWARE_VERSION_REV_POS), data.at(FIRMWARE_VERSION_SUPPORT_POS));
-                }
-                this->get_firmware_version_promise->set_value(response);
-            } else if (command_code == COMMAND_SAM_CONFIGURATION_RESPONSE && this->configure_sam_promise) {
-                this->configure_sam_promise->set_value(true);
-            } else if (command_code == COMMAND_IN_LIST_PASSIVE_TARGET_RESPONSE &&
-                       this->in_list_passive_target_promise) {
-                parseInListPassiveTargetResponse();
-            }
+        if (first_start_code_offset == -1) {
+            // no start code found so far: continue with outer loop and receive more data
+        } else {
+            // use first start code offset found and assume garbage before -> discard this
+            n -= first_start_code_offset;
+            memmove(buf, buf + first_start_code_offset, n);
         }
+    }
+}
 
-    } else {
-        if (this->debug) {
-            EVLOG_warning << "Last byte is NOT 0x00, something went wrong...";
-            std::stringstream data_stream;
-            for (auto& element : data) {
-                data_stream << "0x" << std::setfill('0') << std::setw(2) << std::right << std::hex << (int)element
-                            << " ";
-            }
-            data_stream << "(length: " << std::dec << data.size() << ")";
-
-            EVLOG_info << "Raw data: " << data_stream.str();
+void PN532Serial::parseData() {
+    if (command_code == COMMAND_GET_FIRMWARE_VERSION_RESPONSE && this->get_firmware_version_promise) {
+        PN532Response response;
+        if (data.size() == 4) {
+            EVLOG_debug << "Sending ACK";
+            serialWrite(ACK_FRAME);
+
+            response.valid = true;
+            response.message = std::make_shared<FirmwareVersion>(
+                data.at(FIRMWARE_VERSION_IC_POS), data.at(FIRMWARE_VERSION_VER_POS), data.at(FIRMWARE_VERSION_REV_POS),
+                data.at(FIRMWARE_VERSION_SUPPORT_POS));
+
+            this->get_firmware_version_promise->set_value(response);
+        } else {
+            EVLOG_debug << "Sending NACK";
+            serialWrite(NACK_FRAME);
         }
+    } else if (command_code == COMMAND_SAM_CONFIGURATION_RESPONSE && this->configure_sam_promise) {
+        EVLOG_debug << "Sending ACK";
+        serialWrite(ACK_FRAME);
+        this->configure_sam_promise->set_value(true);
+    } else if (command_code == COMMAND_IN_LIST_PASSIVE_TARGET_RESPONSE && this->in_list_passive_target_promise) {
+        // always send ACK for now
+        EVLOG_debug << "Sending ACK";
+        serialWrite(ACK_FRAME);
+        parseInListPassiveTargetResponse();
     }
 }
 
@@ -248,13 +220,8 @@ void PN532Serial::parseInListPassiveTargetResponse() {
 
                 if (this->debug) {
                     std::stringstream nfcid_stream;
-
-                    for (auto& element : target_data.nfcid) {
-                        nfcid_stream << "0x" << std::setfill('0') << std::setw(2) << std::right << std::hex
-                                     << (int)element << " ";
-                    }
-                    nfcid_stream << "(length: " << std::dec << target_data.nfcid.size() << ")";
-
+                    nfcid_stream << hexdump(target_data.nfcid);
+                    nfcid_stream << " (length: " << std::dec << target_data.nfcid.size() << ")";
                     EVLOG_info << "NFCID: " << nfcid_stream.str();
                 }
 
@@ -270,13 +237,8 @@ void PN532Serial::parseInListPassiveTargetResponse() {
 
                     if (this->debug) {
                         std::stringstream ats_stream;
-
-                        for (auto& element : target_data.ats) {
-                            ats_stream << "0x" << std::setfill('0') << std::setw(2) << std::right << std::hex
-                                       << (int)element << " ";
-                        }
-                        ats_stream << "(length: " << std::dec << target_data.ats.size() << ")";
-
+                        ats_stream << hexdump(target_data.ats);
+                        ats_stream << " (length: " << std::dec << target_data.ats.size() << ")";
                         EVLOG_info << "ATS: " << ats_stream.str();
                     }
                 }
@@ -293,7 +255,7 @@ void PN532Serial::parseInListPassiveTargetResponse() {
 std::future<bool> PN532Serial::configureSAM() {
     this->configure_sam_promise = std::make_unique<std::promise<bool>>();
     this->serialWriteCommand({
-        COMMAND_SAM_CONFIGURATION, // command codee
+        COMMAND_SAM_CONFIGURATION, // command code
         0x01,                      // normal mode
         0x00,                      // no timeout
         0x01                       // P70_IRQ pin driven by PN532
@@ -317,43 +279,51 @@ std::future<PN532Response> PN532Serial::inListPassiveTarget() {
     return this->in_list_passive_target_promise->get_future();
 }
 
-bool PN532Serial::serialWrite(std::vector<uint8_t> data) {
-    write(fd, data.data(), data.size());
+bool PN532Serial::serialWrite(const std::vector<uint8_t>& data) {
+    auto rv = write(fd, data.data(), data.size());
+
+    if (rv != data.size()) {
+        EVLOG_error << "Write failed: " << strerror(errno) << " (" << errno << ")";
+        return false;
+    }
 
     return true;
 }
 
-bool PN532Serial::serialWriteCommand(std::vector<uint8_t> data) {
-    std::vector<uint8_t> data_copy = data;
-    data_copy.push_back(host_to_pn532);
-    uint8_t sum = 0;
-    for (auto element : data_copy) {
-        sum += element;
-    }
-    uint8_t data_length_checksum = (~data_copy.size() & 0xFF) + 0x01;
-    uint8_t inverse = (~sum & 0xFF) + 0x01;
+bool PN532Serial::serialWriteCommand(const std::vector<uint8_t>& data) {
+    // calculate packet length checksum
+    uint8_t lcs = data.size() + 1;
+    lcs = (uint8_t)-lcs;
 
-    if (inverse > 255) {
-        inverse = inverse - 255;
+    // calculate data check sum
+    uint8_t dcs = HOST_TO_PN532;
+    for (auto element : data) {
+        dcs += element;
     }
-
-    std::vector<uint8_t> serial_data = preamble;
-    serial_data.push_back(data_copy.size());
-    serial_data.push_back(data_length_checksum);
-    serial_data.push_back(host_to_pn532);
-
+    dcs = (uint8_t)-dcs;
+
+    std::vector<uint8_t> serial_data;
+    // size: preamble + start code + LEN + LCS + TFI + data + DCS + postamble
+    serial_data.reserve(1 + START_CODE.size() + 1 + 1 + 1 + data.size() + 1 + 1);
+    serial_data.push_back(PREAMBLE);
+    serial_data.insert(serial_data.end(), START_CODE.begin(), START_CODE.end());
+    serial_data.push_back(data.size() + 1);
+    serial_data.push_back(lcs);
+    serial_data.push_back(HOST_TO_PN532);
     serial_data.insert(serial_data.end(), data.begin(), data.end());
-    serial_data.push_back(inverse);
-    serial_data.push_back(postamble);
+    serial_data.push_back(dcs);
+    serial_data.push_back(POSTAMBLE);
+
+    if (this->debug) {
+        EVLOG_info << "Sending bytes: " << hexdump(serial_data);
+    }
 
     return this->serialWrite(serial_data);
 }
 
 bool PN532Serial::reset() {
-    bool success = true;
-    this->serialWrite({0x55, 0x55, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00});
-
-    return success;
+    return this->serialWrite(
+        {0x55, 0x55, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00});
 }
 
 void PN532Serial::enableDebug() {
diff --git a/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.h b/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.h
index 4f1fdaed32dd..3adc56b3f08c 100644
--- a/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.h
+++ b/modules/HardwareDrivers/NfcReaders/PN532TokenProvider/pn532_serial/PN532Serial.h
@@ -29,7 +29,7 @@ struct PN532Response {
 
 struct FirmwareVersion : public PN532Message {
     uint8_t ic;
-    uint8_t ver;     // TODO: also pre-parsed version string of this
+    uint8_t ver; // TODO: also pre-parsed version string of this
     uint8_t rev;
     uint8_t support; // TODO: bool flags
     FirmwareVersion(uint8_t ic, uint8_t ver, uint8_t rev, uint8_t support) :
@@ -65,38 +65,34 @@ struct InListPassiveTargetResponse : public PN532Message {
 class PN532Serial : public Everest::Serial {
 
 public:
-    PN532Serial();
-    ~PN532Serial();
+    PN532Serial() = default;
+    virtual ~PN532Serial() = default;
 
     void readThread();
     void run() override;
     bool reset();
-    bool serialWrite(std::vector<uint8_t> data);
-    bool serialWriteCommand(std::vector<uint8_t> data);
+    bool serialWrite(const std::vector<uint8_t>& data);
+    bool serialWriteCommand(const std::vector<uint8_t>& data);
     std::future<bool> configureSAM();
     std::future<PN532Response> getFirmwareVersion();
     std::future<PN532Response> inListPassiveTarget();
     void enableDebug();
 
 private:
-    std::vector<uint8_t> preamble = {0x00, 0x00, 0xff};
-    uint8_t postamble = 0x00;
-    uint8_t host_to_pn532 = 0xd4;
+    static constexpr uint8_t PREAMBLE = 0x00;
+    static constexpr uint8_t POSTAMBLE = 0x00;
+    static constexpr uint8_t HOST_TO_PN532 = 0xd4;
+    static constexpr uint8_t PN532_TO_HOST = 0xd5;
+    const std::vector<uint8_t> START_CODE{0x00, 0xff};
+    const std::vector<uint8_t> ACK_FRAME{0x00, 0x00, 0xff, 0x00, 0xff, 0x00};
+    const std::vector<uint8_t> NACK_FRAME{0x00, 0x00, 0xff, 0xff, 0x00, 0x00};
     std::unique_ptr<std::promise<bool>> configure_sam_promise;
     std::unique_ptr<std::promise<PN532Response>> get_firmware_version_promise;
     std::unique_ptr<std::promise<PN532Response>> in_list_passive_target_promise;
-    ssize_t start_of_packet = 0;
-    size_t packet_length = 0;
-    bool preamble_start_seen = false;
-    bool preamble_seen = false;
-    bool first_data = true;
-    bool data_length_checksum_valid = false;
-    uint8_t tfi = 0;
     uint8_t command_code = 0;
     std::vector<uint8_t> data;
     bool debug = false;
 
-    void resetDataRead();
     void parseData();
     void parseInListPassiveTargetResponse();
 
-- 
2.34.1

